services-core/src/solution_submission/retry.rs

```diff
@@ -10,170 +12,153 @@ use futures::{
-                // `retry` relies on the gas price always increasing.
-                self.previous_gas_price = self.previous_gas_price.max(gas_estimate);
-            }
-            Err(ref err) => {
-                log::warn!(
-                    "failed to get gas price from gnosis safe gas station: {}",
-                    err
-                );
-            }
-        };
-        self.previous_gas_price
-    }
-}
-
-fn gas_price(estimated_price: f64, price_increase_count: u32, cap: f64) -> f64 {
-    let factor = 1.5f64.powf(price_increase_count as f64);
-    let new_price = estimated_price * factor;
-    cap.min(new_price)
-}
-
+#[derive(Debug)]
 enum FutureOutput {
     SolutionSubmission(Result<(), MethodError>),
     Timeout,
 }
 
-async fn retry(
-    contract: &dyn StableXContract,
-    gas_price_estimating: &(dyn GasPriceEstimating + Sync),
-    async_sleep: &dyn AsyncSleeping,
-    Args {
-        batch_index,
-        solution,
-        claimed_objective_value,
-        gas_price_cap,
-        nonce,
-    }: Args,
-) -> Result<(), MethodError> {
-    let effective_gas_price_cap = (gas_price_cap / MIN_GAS_PRICE_INCREASE_FACTOR).floor();
-    assert!(effective_gas_price_cap <= gas_price_cap);
-    let mut gas_price_estimator =
-        InfallibleGasPriceEstimator::new(gas_price_estimating, DEFAULT_GAS_PRICE);
-    let mut futures = FuturesUnordered::new();
-
-    for gas_price_increase_count in 0u32.. {
-        let estimated_price = gas_price_estimator.estimate().await;
-        let gas_price = gas_price(estimated_price, gas_price_increase_count, gas_price_cap);
-        assert!(gas_price <= gas_price_cap);
+impl<'a> RetryWithGasPriceIncrease<'a> {
+    async fn new_gas_price(&self, args: &Args) -> Option<f64> {
+        let time_remaining = args
+            .target_confirm_time
+            .saturating_duration_since(self.now.instant_now());
+        // TODO: Use real gas limit once the gas estimators take that into account and we can
+        // predict it more accurately.
+        let estimated_gas_price = match self
+            .gas_price_estimating
+            .estimate_with_limits(DEFAULT_GAS_LIMIT, time_remaining)
+            .await
+        {
+            Ok(gas_price) => gas_price,
+            Err(err) => {
+                log::error!("gas estimation failed: {:?}", err);
+                return None;
+            }
+        };
+        let capped_gas_price = estimated_gas_price.min(args.gas_price_cap);
         log::info!(
-            "solution submission try {} with gas price {}",
-            gas_price_increase_count,
-            gas_price
+            "With {} seconds remaining estimated gas price {} capped to {}.",
+            time_remaining.as_secs(),
+            estimated_gas_price,
+            capped_gas_price,
         );
+        Some(capped_gas_price)
+    }
 
-        let solution_submission_future = contract
+    fn submit_solution(&self, args: &Args, gas_price: f64) -> BoxFuture<FutureOutput> {
```

Comment:
These args seem oddly more specifically like `solution_args` or maybe `submission_args`. Just wondering if it might be more clear than just "args"